com.unity.ml-agents.extensions/Runtime/Sensors/GridSensorBase.cs

```diff
@@ -0,0 +1,336 @@
+        /// <param name="numChannelsToAdd"></param>
+        void ChannelsToTexture(int channelIndex, int numChannelsToAdd)
+        {
+            for (int i = 0; i < m_NumCells; i++)
+            {
+                for (int j = 0; j < numChannelsToAdd; j++)
+                {
+                    m_PerceptionColors[i][j] = m_PerceptionBuffer[i * m_CellObservationSize + channelIndex + j];
+                }
+            }
+            m_PerceptionTexture.SetPixels(m_PerceptionColors);
+        }
+
+        /// <summary>
+        /// GetObjectData - returns an array of values that represent the game object
+        /// This is one of the few methods that one may need to override to get their required functionality
+        /// For instance, if one wants specific information about the current gameobject, they can use this method
+        /// to extract it and then return it in an array format.
+        /// </summary>
+        /// <returns>
+        /// A float[] containing the data that holds the representative information of the passed in gameObject
+        /// </returns>
+        /// <param name="currentColliderGo">The game object that was found colliding with a certain cell</param>
+        /// <param name="typeIndex">The index of the type (tag) of the gameObject.
+        ///           (e.g., if this GameObject had the 3rd tag out of 4, type_index would be 2.0f)</param>
+        /// <param name="normalizedDistance">A float between 0 and 1 describing the ratio of
+        ///            the distance currentColliderGo is compared to the edge of the gridsensor</param>
+        /// <example>
+        ///   Here is an example of extenind GetObjectData to include information about a potential Rigidbody:
+        ///   <code>
+        ///     protected override float[] GetObjectData(GameObject currentColliderGo,
+        ///                                     float type_index, float normalized_distance)
+        ///     {
+        ///         float[] channelValues = new float[ChannelDepth.Length]; // ChannelDepth.Length = 4 in this example
+        ///         channelValues[0] = type_index;
+        ///         Rigidbody goRb = currentColliderGo.GetComponent&lt;Rigidbody&gt;();
+        ///         if (goRb != null)
+        ///         {
+        ///             channelValues[1] = goRb.velocity.x;
+        ///             channelValues[2] = goRb.velocity.y;
+        ///             channelValues[3] = goRb.velocity.z;
+        ///         }
+        ///         return channelValues;
+        ///     }
+        ///  </code>
+        /// </example>
+        protected virtual void GetObjectData(GameObject currentColliderGo, int typeIndex, float[] dataBuffer)
+        {
+            dataBuffer[0] = typeIndex;
+        }
+
+        /// <summary>
+        /// Get the observation size for each cell. This will be the size of dataBuffer in GetObjectData().
+        /// </summary>
+        protected virtual int GetObservationSize()
```

Comment:
How about something like `GetCellObservationSize()` to make it clear that it's not the observation size for the whole sensor?